jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/HttpReceiver.java

```diff
@@ -273,148 +217,111 @@ protected void storeCookie(URI uri, HttpField field)
      * This method takes case of notifying {@link org.eclipse.jetty.client.api.Response.HeadersListener}s.
      *
      * @param exchange the HTTP exchange
-     * @return whether the processing should continue
      */
-    protected boolean responseHeaders(HttpExchange exchange)
+    protected void responseHeaders(HttpExchange exchange)
     {
-        if (!updateResponseState(ResponseState.BEGIN, ResponseState.HEADER, ResponseState.TRANSIENT))
-            return false;
-
-        HttpResponse response = exchange.getResponse();
-        if (LOG.isDebugEnabled())
-            LOG.debug("Response headers {}{}{}", response, System.lineSeparator(), response.getHeaders().toString().trim());
-        ResponseNotifier notifier = getHttpDestination().getResponseNotifier();
-        List<Response.ResponseListener> responseListeners = exchange.getConversation().getResponseListeners();
-        notifier.notifyHeaders(responseListeners, response);
-        contentListeners.reset(responseListeners);
-        contentListeners.notifyBeforeContent(response);
-
-        if (!contentListeners.isEmpty())
+        invoker.run(() ->
         {
-            List<String> contentEncodings = response.getHeaders().getCSV(HttpHeader.CONTENT_ENCODING.asString(), false);
-            if (contentEncodings != null && !contentEncodings.isEmpty())
+            if (LOG.isDebugEnabled())
+                LOG.debug("responseHeaders");
+
+            if (exchange.isResponseComplete())
+                return;
+            responseState = ResponseState.HEADERS;
+            HttpResponse response = exchange.getResponse();
+            if (LOG.isDebugEnabled())
+                LOG.debug("Response headers {}{}{}", response, System.lineSeparator(), response.getHeaders().toString().trim());
+            ResponseNotifier notifier = getHttpDestination().getResponseNotifier();
+            List<Response.ResponseListener> responseListeners = exchange.getConversation().getResponseListeners();
+            notifier.notifyHeaders(responseListeners, response);
+
+            if (HttpStatus.isInterim(response.getStatus()))
             {
-                for (ContentDecoder.Factory factory : getHttpDestination().getHttpClient().getContentDecoderFactories())
+                if (LOG.isDebugEnabled())
+                    LOG.debug("Interim response status {}, succeeding", response.getStatus());
+                responseSuccess(exchange, this::receive);
+                return;
+            }
+
+            if (LOG.isDebugEnabled())
+                LOG.debug("Switching to CONTENT state");
+            responseState = ResponseState.CONTENT;
```

Comment:
Can we skip all that follows if `Content-Length == 0`?
Maybe put a TODO for another PR.

Or maybe better, take a `boolean noContent` parameter to `responseHeaders`, because H2 and H3 know immediately if the HEADERS frame is the last, while for H1 and FCGI we may rely on the `Content-Length` field.

Reply:
Added TODO.